postgresql/resource_postgresql_physical_replication_slot.go

```diff
@@ -0,0 +1,81 @@
+package postgresql
+
+import (
+	"database/sql"
+	"fmt"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+func resourcePostgreSQLPhysicalReplicationSlot() *schema.Resource {
+	return &schema.Resource{
+		Create: PGResourceFunc(resourcePostgreSQLPhysicalReplicationSlotCreate),
+		Read:   PGResourceFunc(resourcePostgreSQLPhysicalReplicationSlotRead),
+		Delete: PGResourceFunc(resourcePostgreSQLPhysicalReplicationSlotDelete),
+		Exists: PGResourceExistsFunc(resourcePostgreSQLPhysicalReplicationSlotExists),
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourcePostgreSQLPhysicalReplicationSlotCreate(db *DBConnection, d *schema.ResourceData) error {
+	name := d.Get("name").(string)
+	sql := "SELECT FROM pg_create_physical_replication_slot($1)"
+	if _, err := db.Exec(sql, name); err != nil {
+		return fmt.Errorf("could not create physical ReplicationSlot %s: %w", name, err)
+	}
+	d.SetId(name)
+
+	return nil
+}
+
+func resourcePostgreSQLPhysicalReplicationSlotExists(db *DBConnection, d *schema.ResourceData) (bool, error) {
+	query := "SELECT slot_name FROM pg_catalog.pg_replication_slots WHERE slot_name = $1 and slot_type = 'physical'"
+	var newID string
+	err := db.QueryRow(query, d.Get("name").(string)).Scan(&newID)
+	switch {
+	case err == sql.ErrNoRows:
+		return false, nil
+	case err != nil:
+		return false, err
+	}
+
+	d.SetId(newID)
+	return true, nil
```

Comment:
If we want to be able to `terraform import`, you need to test this based on the resource ID, not the name.

(During an import, this function is called with a `ResourceData` that contains only the ID, `name` will be empty)

```suggestion
	query := "SELECT 1 FROM pg_catalog.pg_replication_slots WHERE slot_name = $1 and slot_type = 'physical'"
	var unused int
	err := db.QueryRow(query, d.Id()).Scan(&unused)
	switch {
	case err == sql.ErrNoRows:
		return false, nil
	case err != nil:
		return false, err

	return true, nil
```

Then in the read function you can fill the name with the ID so it will properly configure the resource.